lib/plugins/aws/provider/awsProvider.js

```diff
@@ -348,39 +365,51 @@ class AwsProvider {
     const stageUpper = this.getStage() ? this.getStage().toUpperCase() : null;
 
     // add specified credentials, overriding with more specific declarations
-    try {
-      impl.addProfileCredentials(result, 'default');
-    } catch (err) {
-      if (err.message !== 'Profile default does not exist') {
-        throw err;
-      }
-    }
-    impl.addCredentials(result, this.serverless.service.provider.credentials); // config creds
-    if (this.serverless.service.provider.profile && !this.options['aws-profile']) {
-      // config profile
-      impl.addProfileCredentials(result, this.serverless.service.provider.profile);
-    }
-    impl.addEnvironmentCredentials(result, 'AWS'); // creds for all stages
-    impl.addEnvironmentProfile(result, 'AWS');
-    impl.addEnvironmentCredentials(result, `AWS_${stageUpper}`); // stage specific creds
-    impl.addEnvironmentProfile(result, `AWS_${stageUpper}`);
-    if (this.options['aws-profile']) {
-      impl.addProfileCredentials(result, this.options['aws-profile']); // CLI option profile
-    }
-    result.region = this.getRegion();
-
-    const deploymentBucketObject = this.serverless.service.provider.deploymentBucketObject;
-    if (
-      deploymentBucketObject &&
-      deploymentBucketObject.serverSideEncryption &&
-      deploymentBucketObject.serverSideEncryption === 'aws:kms'
-    ) {
-      result.signatureVersion = 'v4';
-    }
+    return impl
+      .addProfileCredentials(result, 'default')
+      .catch(err => {
+        if (err.message !== 'Profile default does not exist') {
+          throw err;
+        }
+      })
+      .then(() => {
+        impl.addCredentials(result, this.serverless.service.provider.credentials); // config creds
+        if (this.serverless.service.provider.profile && !this.options['aws-profile']) {
+          // config profile
+          return impl.addProfileCredentials(result, this.serverless.service.provider.profile);
+        }
+        return null;
+      })
+      .then(() => {
+        impl.addEnvironmentCredentials(result, 'AWS'); // creds for all stages
+        return impl.addEnvironmentProfile(result, 'AWS');
+      })
+      .then(() => {
+        impl.addEnvironmentCredentials(result, `AWS_${stageUpper}`); // stage specific creds
+        return impl.addEnvironmentProfile(result, `AWS_${stageUpper}`);
```

Comment:
It's probably a good moment to refactor this logic into something more efficient.

What we're currenlty doing is gathering profile credentials for all possible options, one by one. from one of least priority to top priority. If there are credentials for currently processed option, then it overrides the data in `result` object. So in the end we we have the top priority credential that exists resolved.

What should happen instead is - start with top priority, and stop after first option resolves with credentials, so we do not bother to resolve credentials for profiles which we will not use at all

As this PR now implies a possible async calls for profile creds resolution, then it's quite important we do not make ineffective requests.